File: zeroMQ/cpp/subscriber.cpp
#include <zmq.hpp>
#include <string>
#include <iostream>

int main(int argc, char** argv) {
    if (argc < 3) {
        std::cerr << "Usage: " << argv[0] << " <connect-endpoint> <topic>\n";
        std::cerr << "Example: ./subscriber tcp://localhost:5555 test\n";
        return 1;
    }
    std::string endpoint = argv[1];
    std::string topic = argv[2];

    zmq::context_t ctx{1};
    zmq::socket_t sub(ctx, ZMQ_SUB);
    sub.connect(endpoint);
    sub.setsockopt(ZMQ_SUBSCRIBE, topic.data(), topic.size());

    while (true) {
        zmq::message_t msg;
        auto result = sub.recv(msg); // Using structured bindings
        if (result.has_value()) {  // Check if receive was successful
            std::string data(static_cast<const char*>(msg.data()), msg.size());
            std::cout << "Received: " << data << std::endl;
        }
    }
    return 0;
}

File: zeroMQ/python/subscriber.py
#!/usr/bin/env python3
"""ZeroMQ Python subscriber example
Requires: pip install pyzmq
"""
import sys
import zmq

def main():
    if len(sys.argv) < 3:
        print(f"Usage: {sys.argv[0]} <connect-endpoint> <topic>")
        print("Example: ./subscriber.py tcp://localhost:5555 test")
        return 1
    endpoint = sys.argv[1]
    topic = sys.argv[2]

    ctx = zmq.Context()
    sub = ctx.socket(zmq.SUB)
    sub.connect(endpoint)
    sub.setsockopt_string(zmq.SUBSCRIBE, topic)

    try:
        while True:
            msg = sub.recv_string()
            print(f"Received: {msg}")
    except KeyboardInterrupt:
        pass
    return 0

if __name__ == '__main__':
    sys.exit(main())
